lib/money/allocator.rb

```diff
@@ -132,15 +162,16 @@ def all_rational?(splits)
       splits.all? { |split| split.is_a?(Rational) }
     end
 
-    def allocation_index_for(strategy, length, idx)
-      case strategy
-      when :roundrobin
-        idx % length
-      when :roundrobin_reverse
-        length - (idx % length) - 1
-      else
-        raise ArgumentError, "Invalid strategy. Valid options: :roundrobin, :roundrobin_reverse"
+    # Given a list of decimal numbers, return a list ordered by which is nearest to the next whole number.
+    # For instance, given inputs [1.1, 1.5, 1.9] the correct ranking is 2, 1, 0. This is because 1.9 is nearly 2.
+    # Note that we are not ranking by absolute size, we only care about the distance between our input number and
+    # the next whole number. Similarly, given the input [9.1, 5.5, 3.9] the correct ranking is *still* 2, 1, 0. This
+    # is because 3.9 is nearer to 4 than 9.1 is to 10.
+    def rank_by_nearest(amounts)
+      nearest = amounts.map.with_index  do |amount, i|
+        { :index => i, :distance => 1 - amount[:fractional_subunits] }
       end
+      nearest.sort_by { |n| n[:distance] }.map { |n| n[:index] }
     end
```

Comment:
Actually, according to Phind, it will only iterate over the array once :scream:

https://www.phind.com/search?cache=k6rzfykizt2xp1ohjz6x374u

Reply:
Benchmarked with this code:

```rb
require 'benchmark/ips'

Benchmark.ips do |x|
  # Configure the number of seconds used during
  # the warmup phase (default 2) and calculation phase (default 5)
  x.config(:time => 5, :warmup => 2)

  amounts = Array.new(4) { { whole_subunits: rand(100), fractional_subunits: rand } }


  x.report("one-liner") do
    amounts.each_with_index.sort_by{ |amount, _| 1 - amount[:fractional_subunits] }.map(&:last)
  end

  # Typical mode, runs the block as many times as it can
  x.report("original") do
    nearest = amounts.map.with_index  do |amount, i|
      { :index => i, :distance => 1 - amount[:fractional_subunits] }
    end
    nearest.sort_by { |n| n[:distance] }.map { |n| n[:index] }
  end

  # Compare the iterations per second of the various reports!
  x.compare!
end
```

When there are 4 elements in the array, the original is 1.04x slower:

```
Warming up --------------------------------------
           one-liner    43.185k i/100ms
            original    42.775k i/100ms
Calculating -------------------------------------
           one-liner    440.940k (± 1.5%) i/s -      2.246M in   5.094009s
            original    422.052k (± 2.6%) i/s -      2.139M in   5.071319s

Comparison:
           one-liner:   440940.3 i/s
            original:   422052.3 i/s - 1.04x  slower
```

When there are 1000 elements in the array, the original is 1.28x slower:

```
Warming up --------------------------------------
           one-liner   275.000 i/100ms
            original   217.000 i/100ms
Calculating -------------------------------------
           one-liner      2.802k (± 0.8%) i/s -     14.025k in   5.004871s
            original      2.189k (± 0.9%) i/s -     11.067k in   5.056980s

Comparison:
           one-liner:     2802.4 i/s
            original:     2188.7 i/s - 1.28x  slower
```

So the one-liner that uses the enumerators is fast enough to use the implementation that avoids 3 iterations 🎉